File: 2024/day19/clojure/solution.clj
(ns solution
  (:require [clojure.string :as str]))

(defn parse-input [input-text]
  (let [parts (str/split (str/trim input-text) #"\n\n")
        patterns (->> (str/split (first parts) #",")
                      (map str/trim)
                      vec)
        designs (str/split-lines (second parts))]
    {:patterns patterns
     :designs designs}))

(defn count-ways
  "Count ways to form design from patterns using bottom-up dynamic programming.
   dp[i] = number of ways to form design[i:] from patterns."
  [design patterns]
  (let [n (count design)]
    (-> (reduce
          (fn [dp pos]
            (let [ways (reduce
                         (fn [total pattern]
                           (let [end (+ pos (count pattern))]
                             (if (and (<= end n)
                                      (= pattern (subs design pos end)))
                               (+ total (dp end))
                               total)))
                         0
                         patterns)]
              (assoc dp pos ways)))
          (assoc (vec (repeat (inc n) 0)) n 1)
          (range (dec n) -1 -1))
        (get 0))))

(defn part1 [designs patterns]
  (->> designs
       (filter #(pos? (count-ways % patterns)))
       count))

(defn part2 [designs patterns]
  (transduce (map #(count-ways % patterns)) + designs))

(defn -main [& _args]
  (let [{:keys [patterns designs]} (parse-input (slurp "../input.txt"))]
    (println "Part 1:" (part1 designs patterns))
    (println "Part 2:" (part2 designs patterns))))

(-main)

File: 2024/day19/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Solution {
    public static void main(String[] args) throws IOException {
        var content = Files.readString(Path.of("../input.txt")).trim();
        var parts = content.split("\n\n");

        // Parse patterns (comma-separated on first line)
        var patterns = Arrays.stream(parts[0].split(","))
                             .map(String::trim)
                             .toArray(String[]::new);

        // Parse designs (one per line after blank line)
        var designs = parts[1].trim().split("\n");

        // Part 1: Count designs that can be formed
        var part1 = Arrays.stream(designs)
                          .filter(d -> countWays(d, patterns) > 0)
                          .count();
        System.out.println("Part 1: " + part1);

        // Part 2: Sum the number of ways each design can be formed
        var part2 = Arrays.stream(designs)
                          .mapToLong(d -> countWays(d, patterns))
                          .sum();
        System.out.println("Part 2: " + part2);
    }

    /**
     * Count number of ways to form design from patterns.
     * Uses dynamic programming with memoization.
     */
    private static long countWays(String design, String[] patterns) {
        var memo = new HashMap<Integer, Long>();
        return dpCountWays(design, 0, patterns, memo);
    }

    private static long dpCountWays(String design, int pos, String[] patterns, Map<Integer, Long> memo) {
        if (pos == design.length()) {
            return 1;
        }

        var cached = memo.get(pos);
        if (cached != null) {
            return cached;
        }

        var total = 0L;
        for (var pattern : patterns) {
            var plen = pattern.length();
            if (pos + plen <= design.length() &&
                design.regionMatches(pos, pattern, 0, plen)) {
                total += dpCountWays(design, pos + plen, patterns, memo);
            }
        }

        memo.put(pos, total);
        return total;
    }
}
